controllers/cruisecontroloperation_ttl_controller.go

```diff
@@ -0,0 +1,116 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/banzaicloud/koperator/api/v1alpha1"
+	banzaiv1alpha1 "github.com/banzaicloud/koperator/api/v1alpha1"
+)
+
+const (
+	defaultRequeueCruiseControlOperatonTTL = 30
+)
+
+// CruiseControlOperationTTLReconciler reconciles CruiseControlOperation custom resources
+type CruiseControlOperationTTLReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations,verbs=get;list;watch;create;update;patch;delete;deletecollection
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations/status,verbs=get
+
+//nolint:gocyclo
+func (r *CruiseControlOperationTTLReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+	log.V(1).Info("reconciling CruiseControlOperation custom resources")
+
+	ccOperation := &banzaiv1alpha1.CruiseControlOperation{}
+	if err := r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, ccOperation); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			return reconciled()
+		}
+		// Error reading the object - requeue the request.
+		return requeueWithError(log, err.Error(), err)
+	}
+
+	if ccOperation.GetTTLSecondsAfterFinished() == nil || ccOperation.GetCurrentTask().Finished == nil {
+		return reconciled()
+	}
+
+	operationTTL := time.Duration(*ccOperation.GetTTLSecondsAfterFinished())
+	finishedAt := ccOperation.GetCurrentTask().Finished
+
+	if IsExpired(operationTTL, finishedAt.Time) {
+		log.Info("cleaning up finished CruiseControlOperation", "name", ccOperation.GetName(), "namespace", ccOperation.GetNamespace())
+		return r.delete(ctx, ccOperation)
+	} else {
+		return requeueAfter(int(finishedAt.Time.Add(operationTTL).Sub(time.Now())))
+	}
+}
+
+// SetupCruiseControlWithManager registers cruise control controller to the manager
+func SetupCruiseControlOperationTTLWithManager(mgr ctrl.Manager) *ctrl.Builder {
+	cruiseControlOperationTTLPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newObj := e.ObjectNew.(*banzaiv1alpha1.CruiseControlOperation)
+			if newObj.IsFinished() && newObj.GetTTLSecondsAfterFinished() != nil && newObj.GetDeletionTimestamp().IsZero() {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&banzaiv1alpha1.CruiseControlOperation{}).
+		WithEventFilter(SkipClusterRegistryOwnedResourcePredicate{}).
+		WithEventFilter(cruiseControlOperationTTLPredicate).
+		Named("CruiseControlOperationTTL")
+
+	return builder
+}
+
+func IsExpired(ttl time.Duration, finishedAt time.Time) bool {
+	return finishedAt.Add(ttl).Before(time.Now())
+}
```

Comment:
**Optional**: I **MAY** consider replacing the implementation with `time.Since(finishedAt) < ttl`, because it is easier to read/understand (and that is what I am used to comparing elapsed times/durations), but could be opinionated on my end.